src/main/kotlin/processing/steganography/SteganographyDecoder.kt

```diff
@@ -0,0 +1,108 @@
+package processing.steganography
+
+import javafx.scene.image.PixelReader
+import javafx.scene.image.PixelWriter
+import javafx.scene.image.WritableImage
+import javafx.scene.paint.Color
+import processing.ImageProcessing
+import kotlin.math.ceil
+import kotlin.math.floor
+import kotlin.math.roundToInt
+
+class SteganographyDecoder(private val isDecodeImage: Boolean): ImageProcessing {
+
+    private var result_image: WritableImage? = null
+    private var result_text: String = ""
+
+    override fun process(image: WritableImage) {
+        if (isDecodeImage) {
+            decodeImage(image)
+        } else {
+            decodeText(image)
+        }
+    }
+
+    private fun decodeText(image: WritableImage) {
+        val reader: PixelReader = image.pixelReader
+        var str = ""
+        var length = reader.getArgb(image.width.toInt() - 1, image.height.toInt() - 1) and 0b111111111111111111111111
+        loop@for (x in 0 until image.width.toInt()) {
+            for (y in 0 until image.height.toInt()) {
+                if (length <= 0) break@loop
+                val color = reader.getColor(x, y)
+                val ch1 = ((((color.green * 255).toInt() and 0b1111) shl 4) or ((color.red * 255).toInt() and 0b1111))
+                var res = "${ch1.toChar()}"
```

Comment:
is it the same as ch1.toString()?

Reply:
I have to use ch1.toChar(). Otherwise it's gonna encode "something" as "115o109e116h105n103", which basically turns char into ascii code